network/codec/json/encoder.go

```diff
@@ -17,13 +19,15 @@ type Encoder struct {
 func (e *Encoder) Encode(v interface{}) error {
 
 	// encode the value
-	env, err := encode(v)
+	env, err := v2envEncode(v, "~3net:strm<1")
 	if err != nil {
 		return fmt.Errorf("could not encode value: %w", err)
 	}
 
 	// write the envelope to network
+	p := binstat.EnterTimeVal("~3net:strm<2", "", int64(len(env.Data)))
 	err = e.enc.Encode(env)
+	binstat.Leave(p)
```

Comment:
Overall, I feel these are stylistic preferences. In some places implementations follow the pattern described by Yahya. In other places use use the same pattern Simon used: adding tracing calls directly to the logic. 
There are also downsides of the approach Yahya described in certain cases: it fractures the core logic into many different functions and adds unnecessary levels of indirections which _can_ make the code less legible. 

We generally expect developers to be knowledgeable about a variety of implementation details, because otherwise they can screw up the safety or liveness guarantees of the nodes and thereby undermine network safety as a whole.
Hence, I feel it is not a suitable expectation for this project that developers can be not knowledgeable about aspects of the modules they modify. Rather, we should make it easy for them to understand the aspects they need to.
So the better answer here might be to add a markdown file to `utils/binstat` that explains the details a developer need to know.

A developer anyway needs to understand what is measured in order to decide whether some new code should be included or excluded in the time measurement.

Reply:
@yhassanzadeh13 thanks for the comment!

* There are some issues with the arrangement suggested:
 * The `.Enter*()` functions are still necessary to create `p` used inside the anonymous function.
 * Also, there are many combinations of `.Enter*()` and `.Leave*()` functions which would result in very many `GuageExecutionTime*()` functions.
 * Lastly, the `*Val()` versions of each `.Enter*()` and `.Leave*()` receive a caller value used to create the bin containing the val. If the value is given to `.Enter*()` then it could be given to a `GuageExecutionTime*()` function but if it is to be given to `.Leave*()` then it needs to be somehow calculated after `f()` is executed, and there is no obvious way to do this.
 * Therefore, the compromise is to create a `.Run()` function which does very little but the other functions remain much the same, although they have been changed to take Golang receiver, e.g.:

```
	bs := binstat.EnterTime(fmt.Sprintf("%s%s:%d", via, what, code)) // e.g. 'Time' what is .Run()
	bs.Run(func() {
		data, err = json.Marshal(v) // original expectation more intention revealing
	})
	bs.LeaveVal(int64(len(data))) // e.g. post timing 'Val' form supplying value for bin
```

As you can see the main wish to 'contain' the code being timed to a bin is now within an anonymous function to make it more intention revealing what is going on. I hope this satisfies your comment! Please let me know, as well as if you have more comments. Thanks!

Reply:
> So the better answer here might be to add a markdown file to `utils/binstat` that explains the details a developer need to know.

@AlexHentschel Good news: There is also a markdown file which attempts to be concise :-) In addition there are also test files, one of which uses `binstat` as an external package and tries to illustrate why `binstat` is a useful addition to e.g. `pprof`.

Reply:
Still not sure why we need to create `bs` and perform `LeaveVal` invocation on it externally: 
```
	bs := binstat.EnterTime(fmt.Sprintf("%s%s:%d", via, what, code)) // e.g. 'Time' what is .Run()
	bs.Run(func() {
		data, err = json.Marshal(v) // original expectation more intention revealing
	})
	bs.LeaveVal(int64(len(data))) // e.g. post timing 'Val' form supplying value for bin
```

It seems easy to encapsulate all into a single function: 
```
func GuageExecutionTime(what string, callerParams string, f func()){
     p := enterTimeVal(what string, callerParams string)
     defer leave(p)
     
    f()
}
```

Reply:
@yhassanzadeh13 Yes, it seems easy to combine that particular combination of `.Enter*()` and `.Leave*()` functions into a single function... agreed, but did you read the godoc [1] which describes all the possible types and combinations of `.Enter*()` and `.Leave*()` functions? The possible combinations are:

```
  1. .Enter();...;.Leave()
  2. .EnterVal();...;.Leave()
  3. .EnterTime();...;.Leave()
  4. .EnterTimeVal();...;.Leave()
  5. .Enter();...;.LeaveVal()
  6. .EnterTime();...;.LeaveVal()
```

How does the `GuageExecutionTime()` function know which combination to use?

```
func GuageExecutionTimeVal(what string, val int64, f func()){
    p := enterTimeVal(what, val)
    defer leave(p)
    f()
}
```

Or, if you suggesting to have 6 different versions of the `GuageExecutionTime()` function? In which case, if `.LeaveVal(v)` requires a value v calculated in `f()`, how does that value 'leave' the anonymous function `f()` and become an argument for `.LeaveVal(v)`?

Here is the example in the proposed syntax with the value v determined upon 'leave', but to make it legal code I hard-coded 'data' below:

```
func GuageExecutionTimeValOnLeave(what string, f func()){
    p := enterTime(what)
    defer leaveVal(p, int64(len(data))) // how to make 'data' NOT hard-coded here?
    f()
}

GuageExecutionTimeValOnLeave("...", func() {
	data, err = json.Marshal(v)
})
```

An alternative way might be:

```
var p *BinStat
GuageExecutionTimeValOnLeave("...", &p, func() { // need to store p outside call so that LeaveVal() can use it
	data, err = json.Marshal(v)
	LeaveVal(&p, int64(len(data))) // LeaveVal() can now see BOTH p and data
})
```

But what I came up with is the same number of lines -- 5 -- but arguably less convoluted? And less confusing that `LeaveVal()` is not in the anonymous function?

```
	bs := binstat.EnterTime("...")
	bs.Run(func() {
		data, err = json.Marshal(v)
	})
	bs.LeaveVal(int64(len(data)))
```

Looking forward to your comments. Maybe I missed something? If so, I'm sorry for taking so long on this comment... :-)

[1] https://github.com/onflow/flow-go/blob/simonhf/5589-binstat-efficient-stats-in-bins/utils/binstat/binstat.go#L9

Reply:
Thanks for your detailed explanation 👏 . I appreciate the details. As we talked over the chat, we can take patterns like:
```
// var p *BinStat // we no longer need this
GuageExecutionTimeValOnLeave("...", func() int{ // need to store p outside call so that LeaveVal() can use it
	data, err = json.Marshal(v)
	return len(data)
})

func GaugeExecutionTimeWitLeaveVal(what string, f func() int){
	p := EnterTime("....")
	r := f()
	LeaveVal(&p, r) // LeaveVal() can now see BOTH p and data
}
```

But in the end, as we talked I leave it up to your discretion whether to choose this pattern :).